File: .claude/hooks/lib/transcript.ts
/**
 * Transcript Reader — shared helper for the PreCompact + SessionEnd hooks.
 *
 * Reads Claude Code's per-session jsonl transcript (at
 * ~/.claude/projects/<project-root-slug>/<session-id>.jsonl) and extracts
 * a structured summary: user prompts, assistant text, tool-call counts.
 *
 * Deterministic — no LLM call.
 *
 * Simplified version of the helper used in gavraq/life. For Phase 2 of the
 * AI Training starter, this is all you need. You'll see production-grade
 * versions (with day-scoping, idempotency, etc.) in Gavin's full system.
 */

import { existsSync, readFileSync, readdirSync, writeFileSync, mkdirSync } from 'fs';
import { homedir } from 'os';
import { dirname, join } from 'path';

export interface TranscriptSummary {
  sessionId: string;
  startTime: string | null;
  endTime: string | null;
  userPrompts: string[];
  assistantTexts: string[];
  toolCalls: Record<string, number>;
}

/**
 * Find the jsonl transcript for a session.
 * Tries the project-slug path first; falls back to scanning
 * ~/.claude/projects/* if the hook was invoked from a subdir.
 */
export function findTranscript(sessionId: string, cwd: string): string | null {
  const slug = cwd.replace(/\//g, '-');
  const direct = join(homedir(), '.claude', 'projects', slug, `${sessionId}.jsonl`);
  if (existsSync(direct)) return direct;

  const root = join(homedir(), '.claude', 'projects');
  if (!existsSync(root)) return null;
  for (const dir of readdirSync(root)) {
    const candidate = join(root, dir, `${sessionId}.jsonl`);
    if (existsSync(candidate)) return candidate;
  }
  return null;
}

function isRealPrompt(content: unknown): content is string {
  if (typeof content !== 'string') return false;
  const t = content.trimStart();
  if (t.startsWith('<local-command')) return false;
  if (t.startsWith('<command-name>')) return false;
  if (t.startsWith('<command-message>')) return false;
  if (t.startsWith('<system-reminder>')) return false;
  return t.length >= 2;
}

export function readTranscript(jsonlPath: string): TranscriptSummary {
  const lines = readFileSync(jsonlPath, 'utf-8').split('\n').filter(Boolean);
  const userPrompts: string[] = [];
  const assistantTexts: string[] = [];
  const toolCalls: Record<string, number> = {};
  let sessionId = '';
  let startTime: string | null = null;
  let endTime: string | null = null;

  for (const line of lines) {
    let entry: any;
    try {
      entry = JSON.parse(line);
    } catch {
      continue;
    }
    if (entry.sessionId && !sessionId) sessionId = entry.sessionId;
    if (entry.timestamp) {
      if (!startTime) startTime = entry.timestamp;
      endTime = entry.timestamp;
    }
    if (entry.type === 'user' && isRealPrompt(entry.message?.content)) {
      userPrompts.push(entry.message.content);
    } else if (entry.type === 'assistant' && Array.isArray(entry.message?.content)) {
      for (const block of entry.message.content) {
        if (block?.type === 'text' && typeof block.text === 'string' && block.text.trim()) {
          assistantTexts.push(block.text.trim());
        } else if (block?.type === 'tool_use') {
          const n = block.name || 'unknown';
          toolCalls[n] = (toolCalls[n] || 0) + 1;
        }
      }
    }
  }

  return { sessionId, startTime, endTime, userPrompts, assistantTexts, toolCalls };
}

export function formatSummaryMarkdown(summary: TranscriptSummary, heading: string): string {
  const lines: string[] = ['', heading, ''];
  if (summary.userPrompts.length) {
    lines.push('**Prompts:**');
    for (const p of summary.userPrompts) {
      lines.push(`- ${p.replace(/\s+/g, ' ').trim()}`);
    }
    lines.push('');
  }
  if (summary.assistantTexts.length) {
    lines.push('**Assistant responses:**');
    for (const a of summary.assistantTexts) {
      lines.push(`- ${a.replace(/\s+/g, ' ').trim()}`);
    }
    lines.push('');
  }
  const toolNames = Object.keys(summary.toolCalls).sort();
  if (toolNames.length) {
    lines.push('**Tools:** ' + toolNames.map((n) => `${n}×${summary.toolCalls[n]}`).join(', '));
    lines.push('');
  }
  return lines.join('\n');
}

/**
 * Append a block to a daily log. Creates the file (and parent dirs)
 * if missing. Idempotent on heading — skips if the heading is already present.
 */
export function appendToDailyLog(dailyPath: string, block: string): void {
  mkdirSync(dirname(dailyPath), { recursive: true });
  const existing = existsSync(dailyPath) ? readFileSync(dailyPath, 'utf-8') : '';
  const headingLine = block.split('\n').find((l) => l.startsWith('### '));
  if (headingLine && existing.includes(headingLine)) return;

  const date = dailyPath.split('/').pop()?.replace('.md', '') ?? '';
  const header = existing ? existing : `# ${date}\n`;
  writeFileSync(dailyPath, header + block + '\n', 'utf-8');
}

File: .claude/hooks/lib/transcript.py
"""Transcript Reader - shared helper for the Python variants of
PreCompactFlush + SessionEndSummary hooks.

Mirrors the TypeScript version in transcript.ts. Reads Claude Code's
per-session jsonl transcript (at ~/.claude/projects/<slug>/<session-id>.jsonl)
and extracts a structured summary: user prompts, assistant text, tool-call
counts.

Deterministic - no LLM call.

Simplified version of the helper used in gavraq/life. For Phase 2 of the
AI Training starter, this is all you need. You'll see production-grade
versions (with day-scoping, idempotency, etc.) in Gavin's full system.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path


@dataclass
class TranscriptSummary:
    session_id: str = ""
    start_time: str | None = None
    end_time: str | None = None
    user_prompts: list[str] = field(default_factory=list)
    assistant_texts: list[str] = field(default_factory=list)
    tool_calls: dict[str, int] = field(default_factory=dict)


def find_vault_dir(cwd: Path) -> Path | None:
    """Auto-detect the vault directory inside cwd.

    Looks for any folder named `vault` or ending in `-vault` - matches the
    convention used by LoadMasterPrompt. Returns None if no vault found.
    """
    try:
        for d in sorted(cwd.iterdir()):
            if d.is_dir() and (d.name == "vault" or d.name.endswith("-vault")):
                return d
    except OSError:
        return None
    return None


def find_transcript(session_id: str, cwd: Path) -> Path | None:
    """Find the jsonl transcript for a session.

    Tries the project-slug path first; falls back to scanning
    ~/.claude/projects/* if the hook was invoked from a subdir.
    """
    home = Path.home()
    slug = str(cwd).replace("/", "-")
    direct = home / ".claude" / "projects" / slug / f"{session_id}.jsonl"
    if direct.exists():
        return direct

    root = home / ".claude" / "projects"
    if not root.exists():
        return None
    for sub in root.iterdir():
        candidate = sub / f"{session_id}.jsonl"
        if candidate.exists():
            return candidate
    return None


def _is_real_prompt(content) -> bool:
    """Filter out command / system-reminder noise that shouldn't show in summaries."""
    if not isinstance(content, str):
        return False
    t = content.lstrip()
    if t.startswith(("<local-command", "<command-name>", "<command-message>", "<system-reminder>")):
        return False
    return len(t) >= 2


def read_transcript(jsonl_path: Path) -> TranscriptSummary:
    """Parse a session jsonl into a structured summary."""
    summary = TranscriptSummary()
    try:
        lines = jsonl_path.read_text(encoding="utf-8").splitlines()
    except OSError:
        return summary

    for line in lines:
        if not line.strip():
            continue
        try:
            entry = json.loads(line)
        except (json.JSONDecodeError, ValueError):
            continue

        if entry.get("sessionId") and not summary.session_id:
            summary.session_id = entry["sessionId"]
        ts = entry.get("timestamp")
        if ts:
            if summary.start_time is None:
                summary.start_time = ts
            summary.end_time = ts

        etype = entry.get("type")
        message = entry.get("message", {}) or {}
        content = message.get("content")

        if etype == "user" and _is_real_prompt(content):
            summary.user_prompts.append(content)
        elif etype == "assistant" and isinstance(content, list):
            for block in content:
                if not isinstance(block, dict):
                    continue
                btype = block.get("type")
                if btype == "text" and isinstance(block.get("text"), str) and block["text"].strip():
                    summary.assistant_texts.append(block["text"].strip())
                elif btype == "tool_use":
                    name = block.get("name") or "unknown"
                    summary.tool_calls[name] = summary.tool_calls.get(name, 0) + 1

    return summary


def format_summary_markdown(summary: TranscriptSummary, heading: str) -> str:
    """Render the summary as a markdown block ready to append to a daily log."""
    import re

    lines: list[str] = ["", heading, ""]

    if summary.user_prompts:
        lines.append("**Prompts:**")
        for p in summary.user_prompts:
            collapsed = re.sub(r"\s+", " ", p).strip()
            lines.append(f"- {collapsed}")
        lines.append("")

    if summary.assistant_texts:
        lines.append("**Assistant responses:**")
        for a in summary.assistant_texts:
            collapsed = re.sub(r"\s+", " ", a).strip()
            lines.append(f"- {collapsed}")
        lines.append("")

    tool_names = sorted(summary.tool_calls.keys())
    if tool_names:
        parts = [f"{n}×{summary.tool_calls[n]}" for n in tool_names]
        lines.append("**Tools:** " + ", ".join(parts))
        lines.append("")

    return "\n".join(lines)


def append_to_daily_log(daily_path: Path, block: str) -> None:
    """Append a block to a daily log. Creates the file (and parent dirs)
    if missing. Idempotent on heading - skips if the heading is already present.
    """
    daily_path.parent.mkdir(parents=True, exist_ok=True)
    existing = daily_path.read_text(encoding="utf-8") if daily_path.exists() else ""

    heading_line = next(
        (line for line in block.split("\n") if line.startswith("### ")),
        None,
    )
    if heading_line and heading_line in existing:
        return

    date = daily_path.stem
    header = existing if existing else f"# {date}\n"
    daily_path.write_text(header + block + "\n", encoding="utf-8")
